cache/cache.go

```diff
@@ -13,3 +13,284 @@
+
+// WatchEventStream is the read-only channel returned to a single Watch() caller.
+type WatchEventStream <-chan *clientv3.Event
+
+// Cache buffers a single etcd Watch for a given key‐prefix and fan‑outs local watchers.
+type Cache struct {
+	prefix    string // prefix is the key-prefix this shard is responsible for ("" = root).
+	cfg       Config // immutable runtime configuration
+	client    *clientv3.Client
+	history   *ringBuffer // history stores recent events in a fixed-size ring and supports replay; concrete impl until the API stabilises.
+	demux     *demux      // demux fans incoming events out to active watchers and manages resync.
+	ready     chan struct{}
+	stop      context.CancelFunc
+	waitGroup sync.WaitGroup
+}
+
+// watchCtx collects all the knobs that both serveWatchEvents and watchRetryLoop need.
+type watchCtx struct {
+	client          *clientv3.Client
+	prefix          string
+	demux           *demux
+	ring            *ringBuffer
+	backoffStart    time.Duration
+	backoffMax      time.Duration
+	broadcast       func(*clientv3.Event)
+	onFirstResponse func() // callback to fire once on first upstream response
+}
+
+// New builds a cache shard that watches only the requested prefix.
+// For the root cache pass "".
+func New(client *clientv3.Client, prefix string, opts ...Option) (*Cache, error) {
+	cfg := defaultConfig()
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
+	if cfg.HistoryWindowSize <= 0 {
+		return nil, fmt.Errorf("invalid HistoryWindowSize %d (must be > 0)", cfg.HistoryWindowSize)
+	}
+
+	ringBuffer := newRingBuffer(cfg.HistoryWindowSize)
+
+	internalCtx, cancel := context.WithCancel(context.Background())
+
+	cache := &Cache{
+		prefix:  prefix,
+		cfg:     cfg,
+		client:  client,
+		history: ringBuffer,
+		ready:   make(chan struct{}),
+		stop:    cancel,
+	}
+
+	cache.demux = newDemux(internalCtx, &cache.waitGroup, ringBuffer, cfg.ResyncInterval)
+
+	cache.waitGroup.Add(1)
+	go func() {
+		defer cache.waitGroup.Done()
+		readyOnce := sync.Once{}
+
+		watchCtx := &watchCtx{
+			client:          client,
+			prefix:          prefix,
+			demux:           cache.demux,
+			ring:            cache.history,
+			backoffStart:    cfg.InitialBackoff,
+			backoffMax:      cfg.MaxBackoff,
+			broadcast:       cache.demux.broadcast,
+			onFirstResponse: func() { readyOnce.Do(func() { close(cache.ready) }) },
+		}
+		serveWatchEvents(internalCtx, watchCtx, cfg.UpstreamBufferSize)
+	}()
+
+	return cache, nil
+}
+
+// Watch registers a cache-backed watcher for a given key or prefix.
+// It returns a WatchChan that streams WatchResponses containing events.
+func (c *Cache) Watch(ctx context.Context, key string, opts ...clientv3.OpOption) clientv3.WatchChan {
+	select {
+	case <-c.ready:
+	case <-ctx.Done():
+		emptyWatchChan := make(chan clientv3.WatchResponse)
+		close(emptyWatchChan)
+		return emptyWatchChan
+	}
+
+	if !strings.HasPrefix(key, c.prefix) {
```

Comment:
```suggestion
	if key != c.prefix {
```
Current implementation doesn't support `key != c.prefix`.

Reply:
Will make sure to also update tests to only use exact prefix watches.